### ISSUE TITLE
Do we need to duplicate annotations and builder API?

### ISSUE THREAD

**Issue by rmannibucau** [eclipse/microprofile-open-api#230]
Still thinking out loud a bit but what if at startup an OpenAPI event is fired and the user can customize the OpenAPI instance which is then served directly?

This /2 the number of available API and doesn't really prevent to do anything.

In terms of complexity it stays exactly the same for the end user since annotations are a plain copy of the OpenAPI builder.

**Comment by EricWittmann** [comment]
Are you suggesting removing the both the ModelReader and Filter interfaces and replacing them both with this?

**Comment by rmannibucau** [comment]
No, the proposal was more to drop https://github.com/eclipse/microprofile-open-api/tree/master/api/src/main/java/org/eclipse/microprofile/openapi/annotations and replace it by an event at startup when the application has been discovered (kind of send AnnotatedType metadata and let the user build the OpenAPI with the builder API instead of duplicating that in annotations and have two entry points for end users.

**Comment by EricWittmann** [comment]
But annotations provide the only code-first way of defining your OpenAPI specification.  It's true that there are three ways to produce a document (static file, builder API, annotations) but they each address a different use case.  Removing the annotations eliminates the use-case where developers want to have the OpenAPI spec generated based on the actual JAX-RS code that they write (i.e. code-first development of an API).  I suspect this is the most popular use-case for developers creating JAX-RS APIs.

The builder API is (I suspect) the LEAST popular approach.

**Comment by rmannibucau** [comment]
@EricWittmann not at all, if you do the previous proposal then you can implement it programmatically as well and even control what you modelize. It is not true it breaks jaxrs developpers since the logical lifecycle would be:

1. build openapi model from jaxrs
2. fire the openapi
3. user can customize the model

It is also more efficient for apps because if you add transversally some deployment concerns like security, tracing (to cite only 2 other MP specs so "MUST" be integrated ;)) then the observers can reprocess the openapi to add their own concerns. Today there is no way for a dev or designer (all handled use cases in openapi spec) to do it if the security and tracing is added at deploy time (ops concerns).

In term of code, a code first approach only needs to be able to mark a method to specificy explicitly its operationId, then the matching is pretty trivial and the builder API is perfect for a dev.

In terms of popularity here is the feedback from my fence:

1. colleagues hate annotations cause it is way too verbose (and stereotypes are not mainstream and in this particular case just an indirection so not that great)
2. builder approach is kind of mainstream in other domains thanks to spring java config style which is liked in the industry
3. design driven approach doesn't need that spec at all so no need to do anything (a plain servlet does the work in 10s, no need to invest in anything complex for this kind of users)

So even if the builder is not mainstream it is likely the one:

1. allowing to embrace what MP is
2. allowing to align the programming style on the one most java dev use
3. allowing to get rid of all that boilerplate, the inconsistencies between annotations and builder
4. allowing to get rid of a custom serialization which is not reusable by end users (for persistence, communication between services - client side in particular, etc...) and way too much custom if you try to serialize the OpenAPI model

To summarize: no big issue dropping annotations, it enables spec integrations and it enables the spec to recenter on real world issues IMHO.

Hope the context is cleaner now.

**Comment by arthurdm** [comment]
Sorry, but we're definitely not dropping annotations.  This is the most used form of generating OpenAPI - it goes back to the roots of Swagger.

**Comment by rmannibucau** [comment]
@arthurdm is Microprofile-openapi just a plai copy of swagger adapted to openapi? No hope it fixes the design issues of swagger and embraces CDI programming model?

**Comment by arthurdm** [comment]
@rmannibucau - you do have some valid points, but they must be made as a additional proposal rather than a replacement proposal (as Eric also mentioned).  

So I will close this one, with the hopes that you can articulate your ideas into a additional proposal.  Thx.
